### PR TITLE
Add plastic computations to `material_utility`

### PR DESCRIPTION
**Pull request by ezrayst**
- [x] add dplastic_strain computation
- [x] add pdstrain computation

### CODE REVIEW

File: include/materials/material_utility.tcc [hunk 1]
```diff
@@ -203,3 +203,29 @@ inline const Eigen::Matrix<double, 6, 1> mpm::materials::dtheta_dsigma(
   // Compute and return dtheta / dsigma
   return (dtheta_dr * ((dr_dj2 * dj2_dsigma) + (dr_dj3 * dj3_dsigma)));
 }
+
+//! Compute incremental dplastic_strain
+inline const Eigen::Matrix<double, 6, 1> mpm::materials::dplastic_strain(
```
Comment (bodhinandach) [inline include/materials/material_utility.tcc]: I think this should not be here as different methods have different ways of updating plastic strain. Return mapping algorithm append plastic strain incrementally for example.

File: include/materials/material_utility.tcc [hunk 2]
```diff
@@ -203,3 +203,29 @@ inline const Eigen::Matrix<double, 6, 1> mpm::materials::dtheta_dsigma(
...
+    const Eigen::Matrix<double, 6, 1>& dstress,
+    const Eigen::Matrix<double, 6, 1>& dstrain,
+    const Eigen::Matrix<double, 6, 6>& de) {
+  // Compute and return dplastic_strain
+  return (dstrain - de.inverse() * dstress);
+}
+
+//! Compute plastic deviatoric strain
+inline double mpm::materials::pdstrain(
```
Comment (bodhinandach) [inline include/materials/material_utility.tcc:208]: This is okay though. Each method can call this function to compute pdstrain providing they have the plastic strain.

**Review by bodhinandach — CHANGES_REQUESTED** [review]

File: include/materials/material_utility.tcc (hunk 1 above)
Comment (tianchiTJ) [reply include/materials/material_utility.tcc]: But this is correct no matter what model you use. You can choose to use it or not

**Review by tianchiTJ — APPROVED** [review]

File: include/materials/material_utility.tcc (hunk 1 above)
Comment (bodhinandach) [reply include/materials/material_utility.tcc]: That's true. I am okay with it, but I think we should just put this on the law. Passing all the tensors and matrices to a function and only to perform an inversion is not that efficient I guess...

File: include/materials/material_utility.h [hunk 3]
```diff
@@ -10,59 +10,86 @@ namespace materials {
 
 //! Compute mean stress p (tension positive)
 //! \param[in] stress Stress in Voigt notation where positive is tension
+//! \retval mean stress p (tension positive)
```
Comment (kks32) [inline include/materials/material_utility.h]: `\retval` should be followed by the variable name such as `p` then followed by the explanation

File: include/materials/material_utility.h [hunk 4]
```diff
@@ -10,59 +10,86 @@ namespace materials {
...
 //! \param[in] stress Stress in Voigt notation where positive is tension
 //! \param[in] tolerance Default tolerance value specified by user
+//! \retval derivative of Lode angle theta in terms of stress sigma
 inline const Eigen::Matrix<double, 6, 1> dtheta_dsigma(
     const Eigen::Matrix<double, 6, 1>& stress,
     double tolerance = std::numeric_limits<double>::epsilon());
 
+//! Compute incremental dplastic_strain
+//! \param[in] dstress Incremental stress in Voigt notation where positive is
+//! tension \param[in] dstrain Incremental total strain in Voigt notation where
```
Comment (kks32) [inline include/materials/material_utility.h]: could you please make sure param starts in a new line

File: include/materials/material_utility.tcc (hunk 1 above)
Comment (kks32) [reply include/materials/material_utility.tcc]: I agree with @bodhinandach this might be easier to do locally in the model

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/materials/material_utility.tcc (hunk 2 above)
Comment (ezrayst) [reply include/materials/material_utility.tcc:208]: But this should always be correct right, since even for different methods, dstrain is always divided into dstrain_elastic and dstrain_plastic anyways. The formulation of D stiffness matrix is defined in such a way that we subtract D_elastic. So this is correct. 

It's a hassle though, for one line of code, we have to transfer a lot of variables

File: include/materials/material_utility.h (hunk 4 above)
Comment (ezrayst) [reply include/materials/material_utility.h]: OK

File: include/materials/material_utility.h [hunk 5]
```diff
@@ -10,59 +10,89 @@ namespace materials {
...
 
 //! Compute deviatoric q
 //! \param[in] stress Stress in Voigt notation where positive is tension
+//! \retval q Deviatoric q
 inline double q(const Eigen::Matrix<double, 6, 1>& stress);
 
 //! Compute Lode angle theta (cosine convention)
 //! \param[in] stress Stress in Voigt notation where positive is tension
 //! \param[in] tolerance Default tolerance value specified by user
+//! \retval Lode angle theta (cosine convention)
```
Comment (kks32) [inline include/materials/material_utility.h]: should be a variable name

File: include/materials/material_utility.tcc (hunk 2 above)
Comment (kks32) [reply include/materials/material_utility.tcc:208]: @ezrayst were you replying to the previous comment?

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/materials/material_utility.tcc (hunk 2 above)
Comment (ezrayst) [reply include/materials/material_utility.tcc:208]: Yes, sorry, wrong one

File: include/materials/material_utility.tcc (hunk 1 above)
Comment (ezrayst) [reply include/materials/material_utility.tcc]: But this should always be correct right, since even for different methods, dstrain is always divided into dstrain_elastic and dstrain_plastic anyways. The formulation of D stiffness matrix is defined in such a way that we subtract D_elastic. So this is correct.

It's a hassle though, for one line of code, we have to transfer a lot of variables

Comment (bodhinandach) [reply include/materials/material_utility.tcc]: Yes, exactly, I am not saying that this is wrong. I just feel that this shouldn't be here due to inefficiency and it is easier that each model can perform their own way of computing `dplastic_strain`.

Comment (ezrayst) [reply include/materials/material_utility.tcc]: How about we have this option, and each material can choose whether to use this or not. The same with `j2` and `j3` computations which would not be used too for most models. 

I do feel it is inefficient but so does passing `stress` to only compute `p` right?

Isn't it the purpose of `mpm::materials` to have the common operations? Unless any of you feel that you will never use this at all.

Comment (kks32) [reply include/materials/material_utility.tcc]: Here, you are passing 48 elements to get 6 doubles back, so not sure if we really need this function. Compilers may not inline these functions and that would make it slower. We can have it

**Review by kks32 — CHANGES_REQUESTED** [review]

**Review by kks32 — APPROVED** [review]

File: include/materials/material_utility.tcc (hunk 1 above)
Comment (ezrayst) [reply include/materials/material_utility.tcc]: OK! Waiting for your approval @bodhinandach

Comment (bodhinandach) [reply include/materials/material_utility.tcc]: I am still not sure that this is the best implementation practice. My concern is related to points made by @kks32 and some other points. 

As @kks32 mentioned, this passes 48 elements, 8 times more to compute p, so the scale is a bit different. Imagine if you call this for 1b billion particles. Also think about whether all of those elements can fit into the lowest cache size.

As for `p`, `j2`, and `j3` they are universal and can be used for all types of stresses. However, this specific function only works and correct for models with small-strain elastoplastic assumptions, most likely with explicit implementation. The kinematic decomposition there won't be quite useful for models with large deformation (they use multiplicative decomp), materials with viscous effect (viscous terms are missing), or materials with constrained strains. I think most of the return mapping algorithms also do it incrementally as I mentioned before. And therefore, bring back to my earlier point, why don't we just implement this inside each model?

In conclusion, I can approve this, but this won't give any benefits as compared to computing invariants. There are many limitations in both efficiency and usefulness. I'll let the group decide.

**Review by bodhinandach — APPROVED** [review]

File: include/materials/material_utility.tcc (hunk 1 above)
Comment (ezrayst) [reply include/materials/material_utility.tcc]: A few thoughts:
- I agree this won't work for viscosity, but I think it should work for any elastoplastic materials which are the materials that we usually use anyways. I think it should work for large strains too, and it will give the same results with using `lambda` for instance.
- I agree it's not that efficient with 48 doubles to get one single double. However, coming back to comparison with the computation of p, in NorSand at least, one step of compute stress would require at least 4 calls to compute p, which might be more efficient if we do them within the material model itself.
- My point being, not all functions in `mpm::materials` would be used by all material models. We can choose whether to use this function or not and that is fine because we are doing inline anyways. Didn't we want to make as uniform definitions so it is easier to maintain?

I will wait till tomorrow's discussion before merging this PR.
